File: LV2/2.4/countingstars/countingstars.rb
def check_bound(i, j)
  return true if [i < 0, j < 0, i >= $n, j >= $m].any?
  return true if $graph[i][j] == '#'

  false
end

def dfs(i, j)
  return if check_bound(i, j)

  $graph[i][j] = '#'

  dfs(i + 1, j)
  dfs(i - 1, j)
  dfs(i, j + 1)
  dfs(i, j - 1)
end

ts = 0
while line = gets
  ts += 1
  $n, $m = line.split.map(&:to_i)
  $graph = (1..$n).map { gets.chomp.chars }

  cnt = 0

  $n.times do |i|
    $m.times do |j|
      if $graph[i][j] == '-'
        dfs(i, j)
        cnt += 1
      end
    end
  end
  puts "Case #{ts}: #{cnt}"
end

File: LV2/2.4/countingstars/countingstars.py
import sys
sys.setrecursionlimit(100000)


def dfs(i, j):
    if any([i < 0, j < 0, i >= n, j >= m]):
        return
    if graph[i][j] == '#':
        return

    graph[i][j] = '#'

    dx = [1, -1, 0, 0]
    dy = [0, 0, 1, -1]

    for d in range(4):
        dfs(i+dx[d], j+dy[d])


ts = 0
while True:
    try:
        ts += 1
        n, m = [int(i) for i in input().split(' ')]
        graph = [list(input()) for _ in range(n)]

        cnt = 0
        for i in range(n):
            for j in range(m):
                if graph[i][j] == '-':
                    dfs(i, j)
                    cnt += 1
        print(f"Case {ts}: {cnt}")
    except EOFError:
        break
